### PR TITLE
CustomSelectControl: improve props type inferring

### PR DESCRIPTION
**Pull request by meteorlxy**
## What?

Improve props type inferring for `CustomSelectControl` component.

## Why?

The current implementation (introduced in #63985) has some drawbacks:

1. The option type could be inferred from `value` and `onChange`, which is not expected.
2. The `options` does not accept immutable array.

## How?

Followed the implementation of #64069 :

1. Make use of `NoInfer`
2. Use `ReadonlyArray` for `options`

cc @mirka 

## Testing Instructions

Added type checking in unit tests

### Testing Instructions for Keyboard

N/A

## Screenshots or screencast 

N/A

### CODE REVIEW

File: packages/components/src/custom-select-control/test/index.tsx [hunk 1]
```diff
@@ -689,3 +690,127 @@ describe.each( [
...
+		const onChange: (
+			changeObject: CustomSelectChangeObject< CustomSelectOption >
+		) => void = () => {};
+		const onChangeWithFoo: (
+			changeObject: CustomSelectChangeObject<
+				CustomSelectOption & {
+					foo: string;
+				}
+			>
+		) => void = () => {};
```
Comment (mirka) [inline packages/components/src/custom-select-control/test/index.tsx]: Ideally we would write these tests without having to import the `CustomSelectChangeObject` and `CustomSelectOption` types directly. Consumers outside of this package don't have access to these internal types.

```suggestion
		const onChange: React.ComponentProps<
			typeof UncontrolledCustomSelectControl
		>[ 'onChange' ] = () => {};
```

And for the `onChangeWithFoo` cases, I think we could simplify it into a more realistic example and just inline it:

```tsx
		<UncontrolledCustomSelectControl
			label="Label"
			options={ options }
			value={ {
				key: 'narrow',
				name: 'Narrow',
			} }
			// @ts-expect-error: the option type should not be inferred from `onChange`
			onChange={ ( obj ) => obj.foo }
		/>;
```

File: packages/components/src/custom-select-control/types.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ export type CustomSelectOption = {
 /**
  * The object returned from the onChange event.
  */
-type CustomSelectChangeObject< T extends CustomSelectOption > = {
+export type CustomSelectChangeObject< T extends CustomSelectOption > = {
```
Comment (mirka) [inline packages/components/src/custom-select-control/types.ts]: Better if we could remove this export, as noted above.

**Review by mirka — COMMENTED** [review]
Thanks for the improvement!

File: packages/components/src/custom-select-control/types.ts (hunk 2 above)
Comment (meteorlxy) [reply packages/components/src/custom-select-control/types.ts]: Removed

File: packages/components/src/custom-select-control/test/index.tsx (hunk 1 above)
Comment (meteorlxy) [reply packages/components/src/custom-select-control/test/index.tsx]: Updated.

As for the `onChangeWithFoo` cases, what you provided above should not be what we want.

We need to test if the `NoInfer` in the `onChange` method works as expected, so we have to define the type of the `onChange` prop explicitly.

It should not be a realistic use case for sure, but we still need to test it. So I added some comments to explain it.

File: packages/components/src/custom-select-control/test/index.tsx [hunk 3]
```diff
@@ -689,3 +689,128 @@ describe.each( [
 		} );
 	} );
 } );
+
+describe( 'Type checking', () => {
```
Comment (tyxla) [inline packages/components/src/custom-select-control/test/index.tsx:693]: Not something for this PR, but I wonder if we should start putting these type-checking pseudo-tests in a separate file under `/test` (types.tsx for example).

Then we could add a global ESLint disable rule for `jest/expect-expect` and we'll no longer have to do it every time. 

Thoughts?

Comment (tyxla) [reply packages/components/src/custom-select-control/test/index.tsx:693]: cc @mirka and @DaniGuardiola

Comment (mirka) [reply packages/components/src/custom-select-control/test/index.tsx:693]: Extracted to #64588

File: packages/components/src/custom-select-control/test/index.tsx (hunk 1 above)
Comment (mirka) [reply packages/components/src/custom-select-control/test/index.tsx]: Ah, you're right. Thanks for the simplification and code comment though, it reads a lot better 👍

**Review by mirka — APPROVED** [review]
Thank you for the nice improvement 🚀

**Pull request merged**